[Skia] Build a SkTextBlob when recording DrawGlyphs display list item

carlosgcampos opened the pull request:
#### 7e98db0c73bf24cbd67d25ae31629cb4f3967607
<pre>
[Skia] Build a SkTextBlob when recording DrawGlyphs display list item
<a href="https://bugs.webkit.org/show_bug.cgi?id=299869">https://bugs.webkit.org/show_bug.cgi?id=299869</a>

Reviewed by Nikolas Zimmermann.

GlyphDisplayListCache uses a display list to cache the shaping results
of a text run that is frequently painted, but we are still building a
SkTextBlob for the same shaping results every time the display list is
replayed, which is expensive. We can build the SkTextBlob once when
recording and save that in DrawGlyphs item instead of the list of glyphs
and advances.

Canonical link: <a href="https://commits.webkit.org/300818@main">https://commits.webkit.org/300818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a2fbc12f636e08f41ea2a1fa04e99f7e1f3e2b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123954 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130745 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52263 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94283 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110883 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74884 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34319 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29045 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74217 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29267 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133437 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50907 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38775 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102750 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51281 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102574 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26094 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47921 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26174 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47760 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50760 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56524 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50234 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53580 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
```diff
@@ -276,15 +280,20 @@ void DrawFilteredImageBuffer::dump(TextStream& ts, OptionSet<AsTextFlag> flags)
+    if (m_textBlob)
+        static_cast<GraphicsContextSkia*>(&context)->drawSkiaText(m_textBlob, SkFloatToScalar(m_localAnchor.x()), SkFloatToScalar(m_localAnchor.y()), m_enableAntialiasing, m_isVertical);
+#else
     return context.drawGlyphs(m_font, m_glyphs.span(), m_advances.span(), m_localAnchor, m_fontSmoothingMode);
```
The return is superfluous and should be removed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
```diff
@@ -276,15 +280,20 @@ void DrawFilteredImageBuffer::dump(TextStream& ts, OptionSet<AsTextFlag> flags)
 {
+#if USE(SKIA)
+    if (m_textBlob)
+        static_cast<GraphicsContextSkia*>(&context)->drawSkiaText(m_textBlob, SkFloatToScalar(m_localAnchor.x()), SkFloatToScalar(m_localAnchor.y()), m_enableAntialiasing, m_isVertical);
```
Hmm, what about other types of GraphicsContexts also used by our ports? e.g. NullGraphicsContext used as GraphicsContext? I think this is error-prone. Why not introduce a new cross-platform virtual method, that only our ports implement?

nikolaszimmermann commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
```diff
@@ -495,7 +495,28 @@ class DrawGlyphs {
+        : m_localAnchor(localAnchor)
+        , m_fontSmoothingMode(smoothingMode)
+    {
+        m_textBlob = font->buildTextBlob(glyphs, advances, smoothingMode, m_enableAntialiasing, m_isVertical);
```
This can go into the initializer list as well.

carlosgcampos replied:
Turns out we can't do that. My guess is that m_enableAntialiasing and m_isVertical are first set by buildTextBlob and after that set to their default value because of they are after m_textBlob in the class definition.

carlosgcampos commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
```diff
@@ -276,15 +280,20 @@ void DrawFilteredImageBuffer::dump(TextStream& ts, OptionSet<AsTextFlag> flags)
 void DrawGlyphs::apply(GraphicsContext& context) const
 {
+#if USE(SKIA)
+    if (m_textBlob)
```
Yes, if font size is 0, for example.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fd3f917a23415d6fa89c4788b31489a6d5d9f264. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51557

nikolaszimmermann commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -52,6 +52,12 @@
 
+#if USE(SKIA)
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN
+#include <skia/core/SkTextBlob.h>
```
We shouldn't pull in Skia headers here if avoidable -- since SkTextBlob isn't stored as member, a `class SkTextBlob` should be sufficient, no?

nikolaszimmermann commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -241,6 +247,10 @@ class Font : public RefCounted<Font>, public CanMakeSingleThreadWeakPtr<Font> {
 #endif
 
+#if USE(SKIA)
+    sk_sp<SkTextBlob> buildTextBlob(std::span<const GlyphBufferGlyph>, std::span<const GlyphBufferAdvance>, FontSmoothingMode, bool& enableAntialiasing, bool& isVertical) const;
```
I didn't realize before this was passing those bools as references...

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
```diff
@@ -30,74 +30,17 @@
-
-    auto blob = builder.make();
-    static_cast<GraphicsContextSkia*>(&graphicsContext)->drawSkiaText(blob, SkFloatToScalar(position.x()), SkFloatToScalar(position.y()), edging != SkFont::Edging::kAlias, isVertical);
+    bool enableAntialiasing, isVertical;
```
We only ever use one variable per line. And I'd strongly advocate to initialize them with `{ false }` - to not depend on whatever `buildTextBlob` does.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/FontSkia.cpp:
```diff
@@ -197,4 +198,63 @@ bool Font::platformSupportsCodePoint(char32_t character, std::optional<char32_t>
     return m_platformData.skFont().getTypeface()->unicharToGlyph(character);
 }
 
+sk_sp<SkTextBlob> Font::buildTextBlob(std::span<const GlyphBufferGlyph> glyphs, std::span<const GlyphBufferAdvance> advances, FontSmoothingMode smoothingMode, bool& enableAntialias, bool& isVertical) const
```
I don't like that enableAntialias and isVertical are passed in as references, and only conditionally set. If m_platformData.size() or glyphs.size() is empty, they are not set at all.

I'd rather introduce a specific `computeEnableAntialias()` method and `computeIsVerticalText()` and use them at the calcite to initialize m_enableAntialias, m_isVertical, and then pass those bools here by value, not by reference.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1a2fbc12f636e08f41ea2a1fa04e99f7e1f3e2b3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51557

nikolaszimmermann commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -241,6 +247,11 @@ class Font : public RefCounted<Font>, public CanMakeSingleThreadWeakPtr<Font> {
 
+#if USE(SKIA)
+    sk_sp<SkTextBlob> buildTextBlob(std::span<const GlyphBufferGlyph>, std::span<const GlyphBufferAdvance>, FontSmoothingMode) const;
+    bool enableAntialiasing(FontSmoothingMode) const;
```
We usually use prefixes like "use" / "can" / "should" for boolean conditions like this: "bool shouldEnableAntialiasing".

carlosgcampos replied:
I'm not sure we are very consistent, see `FontCascade::enableKerning()` for example...

nikolaszimmermann reviewed (approved):
Looks great, thanks @carlosgcampos!

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontSkia.cpp:
```diff
@@ -197,4 +198,64 @@ bool Font::platformSupportsCodePoint(char32_t character, std::optional<char32_t>
+        return SkFont::Edging::kAlias;
+    }
+
+    RELEASE_ASSERT_NOT_REACHED();
```
No, if it's a release assert.

carlosgcampos committed:
[Skia] Build a SkTextBlob when recording DrawGlyphs display list item
https://bugs.webkit.org/show_bug.cgi?id=299869

Reviewed by Nikolas Zimmermann.

GlyphDisplayListCache uses a display list to cache the shaping results
of a text run that is frequently painted, but we are still building a
SkTextBlob for the same shaping results every time the display list is
replayed, which is expensive. We can build the SkTextBlob once when
recording and save that in DrawGlyphs item instead of the list of glyphs
and advances.

Canonical link: https://commits.webkit.org/300818@main
